File: common/src/main/java/juuxel/adorn/block/ChimneyBlock.java
package juuxel.adorn.block;

import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.entity.player.Player;
import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.InteractionResult;
import net.minecraft.util.StringRepresentable;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.core.BlockPos;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.Level;

public final class ChimneyBlock extends AbstractChimneyBlock implements BlockWithDescription {
    public static final EnumProperty<SmokeType> SMOKE_TYPE = EnumProperty.create("smoke_type", SmokeType.class);
    private static final String DESCRIPTION_KEY = "block.adorn.chimney.description";

    public ChimneyBlock(BlockBehaviour.Properties settings) {
        super(settings);
        registerDefaultState(defaultBlockState().setValue(SMOKE_TYPE, SmokeType.CAMPFIRE));
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(SMOKE_TYPE);
    }

    @Override
    protected InteractionResult useWithoutItem(BlockState state, Level world, BlockPos pos, Player player, BlockHitResult hit) {
        world.setBlockAndUpdate(pos, state.cycle(SMOKE_TYPE));
        return InteractionResult.SUCCESS;
    }

    @Override
    public void animateTick(BlockState state, Level world, BlockPos pos, RandomSource random) {
        if (state.getValue(CONNECTED)) return;

        int count = 3 + random.nextInt(2);
        switch (state.getValue(SMOKE_TYPE)) {
            case CLASSIC -> {
                for (int i = 0; i < count; i++) {
                    world.addAlwaysVisibleParticle(
                        ParticleTypes.LARGE_SMOKE,
                        pos.getX() + 0.3 + random.nextDouble() * 0.4,
                        pos.getY() + 0.9,
                        pos.getZ() + 0.3 + random.nextDouble() * 0.4,
                        0.0, 0.0, 0.0
                    );
                }
            }

            case CAMPFIRE -> {
                for (int i = 0; i < count; i++) {
                    world.addAlwaysVisibleParticle(
                        ParticleTypes.CAMPFIRE_COSY_SMOKE, true,
                        pos.getX() + 0.3 + random.nextDouble() * 0.4,
                        pos.getY() + 0.9 + random.nextDouble(),
                        pos.getZ() + 0.3 + random.nextDouble() * 0.4,
                        0.0, 0.07, 0.0
                    );
                }
            }
        }
    }

    public enum SmokeType implements StringRepresentable {
        CLASSIC("classic"), CAMPFIRE("campfire");

        private final String id;

        SmokeType(String id) {
            this.id = id;
        }

        @Override
        public String getSerializedName() {
            return id;
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/ChimneyBlock.kt
@file:Suppress("DEPRECATION")
package juuxel.adorn.block

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.Blocks
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.particle.ParticleTypes
import net.minecraft.state.StateManager
import net.minecraft.state.property.EnumProperty
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.StringIdentifiable
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import java.util.Random

class ChimneyBlock : AbstractChimneyBlock(Settings.copy(Blocks.BRICKS).ticksRandomly().nonOpaque()), BlockWithDescription {
    override val descriptionKey = "block.adorn.chimney.desc"

    init {
        defaultState = defaultState.with(SMOKE_TYPE, SmokeType.CAMPFIRE)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(SMOKE_TYPE)
    }

    override fun onUse(
        state: BlockState, world: World, pos: BlockPos, player: PlayerEntity?, hand: Hand?, hitResult: BlockHitResult?
    ): ActionResult {
        world.setBlockState(pos, state.cycle(SMOKE_TYPE))
        return ActionResult.SUCCESS
    }

    @Environment(EnvType.CLIENT)
    override fun randomDisplayTick(state: BlockState, world: World, pos: BlockPos, random: Random) {
        if (state[CONNECTED]) return

        when (state[SMOKE_TYPE]!!) {
            SmokeType.CLASSIC -> for (i in 1..(3 + random.nextInt(2))) {
                world.addImportantParticle(
                    ParticleTypes.LARGE_SMOKE,
                    pos.x + 0.3 + random.nextDouble() * 0.4,
                    pos.y + 0.9,
                    pos.z + 0.3 + random.nextDouble() * 0.4,
                    0.0, 0.0, 0.0
                )
            }

            SmokeType.CAMPFIRE -> for (i in 1..(3 + random.nextInt(2))) {
                world.addImportantParticle(
                    ParticleTypes.CAMPFIRE_COSY_SMOKE, true,
                    pos.x + 0.3 + random.nextDouble() * 0.4,
                    pos.y + 0.9 + random.nextDouble(),
                    pos.z + 0.3 + random.nextDouble() * 0.4,
                    0.0, 0.07, 0.0
                )
            }
        }
    }

    companion object {
        val SMOKE_TYPE = EnumProperty.of("smoke_type", SmokeType::class.java)
    }

    enum class SmokeType(private val id: String) : StringIdentifiable {
        CLASSIC("classic"), CAMPFIRE("campfire");

        override fun asString() = id
    }
}
